src/darwin/Framework/CHIP/MTRDeviceControllerDataStore.mm

```diff
@@ -273,6 +275,320 @@ - (nullable MTRCASESessionResumptionInfo *)_findResumptionInfoWithKey:(nullable
     return resumptionInfo;
 }
 
+#pragma - Attribute Cache utility
+
+/** MTRDevice cache storage
+ *
+ *  Per controller:
+ *    Last serial number
+ *        key: "last serial number"
+ *        value: uint64 serial number
+ *    NodeID index
+ *        key: "attrCacheNodeIndex"
+ *        value: list of nodeIDs
+ *    EndpointID index
+ *        key: "attrCacheEndpointIndex:<nodeID>:endpointID"
+ *        value: list of endpoint IDs
+ *    ClusterID index
+ *        key: "<nodeID+endpointID> clusters"
+ *        value: list of cluster IDs
+ *    AttributeID index
+ *        key: "<nodeID+endpointID+clusterID> attributes"
+ *        value: list of attribute IDs
+ *    Attribute data entry:
+ *        key: "<nodeID+endpointID+clusterID+attributeID> attribute data"
+ *        value: serialized dictionary of attribute data
+ *
+ *    Attribute data dictionary
+ *        Additional value "serial number"
+ */
+
+- (id)_fetchAttributeCacheValueForKey:(NSString *)key expectedClass:(Class)expectedClass;
+{
+    __block id data;
+    dispatch_sync(_storageDelegateQueue, ^{
```

Comment:
Storing can definitely be made async, and I'll do that in this PR.

Initializing the MTRDevice with persisted values is a little trickier because I'd need to invent some new mechanism to have MTRDevice wait for the load to finish.

Filed issue for read: https://github.com/project-chip/connectedhomeip/issues/31965